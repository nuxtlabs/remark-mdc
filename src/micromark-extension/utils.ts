import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
