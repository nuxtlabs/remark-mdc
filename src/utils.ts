export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
