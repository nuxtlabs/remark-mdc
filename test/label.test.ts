import { describe } from 'vitest'
import { runMarkdownTests } from './utils'

describe('label', () => {
  runMarkdownTests({
    simple: {
      markdown: ':test[label]'
    },
    empty: {
      markdown: ':test[]',
      expected: ':test'
    },
    'scaped-characters': {
      markdown: ':test[scape \\[ character]'
    },
    'invalid-label-eol': {
      markdown: ':test[\n',
      expected: '\\:test\\['
    },
    'invalid-label-eof': {
      markdown: ':test[',
      expected: '\\:test\\['
    }
  })
})
