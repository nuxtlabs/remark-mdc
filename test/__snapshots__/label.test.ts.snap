// Vitest Snapshot v1

exports[`label > empty 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 8,
          "line": 1,
          "offset": 7,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "textComponent",
    },
  ],
  "position": {
    "end": {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`label > invalid-label-eof 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": ":test[",
        },
      ],
      "position": {
        "end": {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`label > invalid-label-eol 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": ":test[",
        },
      ],
      "position": {
        "end": {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 2,
      "offset": 7,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`label > scaped-characters 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "position": {
            "end": {
              "column": 25,
              "line": 1,
              "offset": 24,
            },
            "start": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
          },
          "type": "text",
          "value": "scape [ character",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 26,
          "line": 1,
          "offset": 25,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "textComponent",
    },
  ],
  "position": {
    "end": {
      "column": 26,
      "line": 1,
      "offset": 25,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`label > simple 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
          },
          "type": "text",
          "value": "label",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "textComponent",
    },
  ],
  "position": {
    "end": {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
